File: 8.composite/composite.ts
// Componente base
interface FileSystem {
    showDetails(): void;
}

// Hoja
class File implements FileSystem {
    constructor(private name: string) {}

    public showDetails(): void {
        console.log(`Archivo: ${this.name}`);
    }
}

// Compuesto
class Directory implements FileSystem {
    private children: FileSystem[] = [];

    constructor(private name: string) {}

    public add(component: FileSystem): void {
        this.children.push(component);
    }

    public showDetails(): void {
        console.log(`Directorio: ${this.name}`);
        this.children.forEach(child => child.showDetails());
    }
}

// Uso
const file1 = new File("archivo1.txt");
const file2 = new File("archivo2.txt");

const directory = new Directory("Documentos");
directory.add(file1);
directory.add(file2);

directory.showDetails();

File: 8.composite/composite.py
from abc import ABC, abstractmethod

# Componente base
class FileSystem(ABC):
    @abstractmethod
    def show_details(self):
        pass

# Hoja
class File(FileSystem):
    def __init__(self, name):
        self.name = name

    def show_details(self):
        print(f"Archivo: {self.name}")

# Compuesto
class Directory(FileSystem):
    def __init__(self, name):
        self.name = name
        self.children = []

    def add(self, component: FileSystem):
        self.children.append(component)

    def show_details(self):
        print(f"Directorio: {self.name}")
        for child in self.children:
            child.show_details()

# Uso
file1 = File("archivo1.txt")
file2 = File("archivo2.txt")

directory = Directory("Documentos")
directory.add(file1)
directory.add(file2)

directory.show_details()
